### PR TITLE
makes sense of implicit macros!

### PR DESCRIPTION
**Pull request by xeno-by**
Supersedes https://github.com/scala/scala/pull/2464

review @paulp @retronym

### CODE REVIEW

**Comment by xeno-by** [comment]
@heathermiller @phaller @milessabin @mandubian @non @tixxit @larsrh please verify that this pull request meets your requirements for macro-based type class materializers

**Comment by paulp** [comment]
What is the meaning of all the nomasters? This isn't all being backported I take it.

**Comment by xeno-by** [comment]
Master has an incomplete and incompatible fix to SI-5923 of its own. Also SI-7291 is indeed backported. As almost everything in the PR depends on either of those, it will be difficult to merge it into master as is. I'll submit a follow-up PR once this one gets in.

**Comment by xeno-by** [comment]
The failures are my fault. Will fix.

**Comment by xeno-by** [comment]
The failures were the fault of the not entirely correct hack put in place to ensure backward compatibility with SI-5353. Now should be fixed.

**Comment by xeno-by** [comment]
@paulp My attempts to emulate current behavior of SI-5353 for the sake of ultimate backward compatibility haven't been very successful, so once again I propose to just disable `neg/t5353.scala` in 2.10.x. To quote the commit message for motivation:

```
[nomaster] temporarily breaks SI-5353

The subsequent fix to SI-5923 unmasks the fact that SI-5353 has not
been fixed - it's just that one of its manifestation got hidden behing
SI-5923. In fact, if in the code snippet from the bug report we change
Array() to Array[Nothing](), things will start crashing as usual.

The problem we have here is that arrays of nothings and nulls are very
weird in a sense that their compile-time representations (types) and
their runtime representations (JVM arrays of Object) behave differently
with respect to subtyping. Due to an unlucky coincidence SI-5923 prevented
some of the arrays of nothing from being compilable, so the problem was
well hidden until now.

A principled approach to handling the situation we have here would be to
fix SI-5353 (we already know how to do that: https://github.com/scala/scala/pull/2486)
and to disallow arrays of nothings and nulls as suggested in SI-7453.
Unfortunately, both fixes are going to bring incompatibilities, which are
not acceptable in a minor release (this pull request targets 2.10.x).

Therefore we decided to turn a blind eye on the iceberg and just fix a tip
of it by emulating the portion of SI-5923 that used to mask SI-5353, retaining
perfect backward compatibility.

Unfortunately, it's not that easy. Apparently one cannot simply report
all the occurrences of Array() as errors, because if we know expected type
of that expression, then everything's fine - the expected type will drive type
inference and the dreaded Array[Nothing]() will not arise. All right,
so let's just check whether pt == WildcardType and then report errors.
However that's still not enough because of SI-3859.

Having my hack failing for the third time, made me stop for a second and
think whether it's worth to play with fire and introduce potential
breakages for the sake of preventing a quite unlikely bug from happening.
I don't think the it's okay to risk here, therefore I just disable the
failing test, especially because we already have a working fix to SI-5353
submitted to master, so it's not like we're deferring the work to be done
to a random point in unclear future.

NOTE: That's only a temporary hack targeted at 2.10.x. There's no
reason for this code to be merged into master, because master is soon
going to have a principled solution to the problem:
https://github.com/scala/scala/pull/2486.
```

**Comment by xeno-by** [comment]
Apparently introducing methods in `scala.reflect.internal` isn't binary compatible. Okay, moving the offending method away from `Symbols.scala`.

**Comment by retronym** [comment]
You are allowed to add exceptions for that package. Unfortunately wildcards aren't supported.

**Comment by xeno-by** [comment]
Hmm, okay. Let's have it pass all the tests first, and then we can think about prettification.

**Comment by tixxit** [comment]
Well, I just tried it out with Spire's Auto type classes. Had to adjust the macros to not take any argument lists and it worked great. This will definitely be quite the feature, thanks!

**Comment by xeno-by** [comment]
@tixxit Thank you for taking a look!

**Comment by mandubian** [comment]
I'll try to test it with Play's macro on next week. In those macros, we mainly infer simple typeclass implicits...

BTW, I've also seen a few tests showing some potential improvements in resolving implicit requiring implicits and this might be cool ;)

Thks for hard work

**Comment by xeno-by** [comment]
Oh man, almost forgot. review @hubertp

**Comment by milessabin** [comment]
LGTM ... awesome work ... many thanks :-)

**Comment by larsrh** [comment]
+1 from me. Works like a charm for shapeless-contrib. Thank you!

**Comment by hubertp** [comment]
Handling diverging implicit expansions lgtm.

**Comment by xeno-by** [comment]
@all Unfortunately we had a disagreement about whether it makes sense to allow implicit macros to affect type inference (as in synthesize Iso's). Therefore I've moved the controversial bits into a separate pull request: https://github.com/scala/scala/pull/2499. Sorry, @milessabin

**Comment by milessabin** [comment]
What was the nature of the controversy?

**Comment by xeno-by** [comment]
@milessabin Let's switch to the discussion of the follow-up pull request to proceed. I'll try to get @odersky to elaborate on that.

**Comment by xeno-by** [comment]
Some feedback from Jason:
1) Provide a test for the `macroExpandAll` commit.
2) In the `topic/iso` pull request explain why `context.undetparams` isn't good enough, and why we have to scan `expectedTpe` for undets.
3) `case OpenImplicit(...) => ...` in the fix to Si-7261 could be simplified.

**Comment by xeno-by** [comment]
@retronym et al. Could we please have a review before the 2.10.2-RC1 deadline?

**Comment by gkossakowski** [comment]
This cannot be merged now. Also, do we think merging this right before RC1 is a good idea? I'm asking reviewers.

**Comment by xeno-by** [comment]
Can be merged now.

**Comment by xeno-by** [comment]
Guys, before we exclude this pull request from our 2.10.2 agenda, please note its significant value.

On the one hand, this PR fixes an important use case for macros, which will immediately improve user experience for Play, Spire, scalaz and shapeless among others. 

On the other hand, it is the key enabler for scala-pickling, which we've managed to successfully port to 2.10.x, solely thanks to this pull request. If this PR doesn't get in by 2.10.2, we will have to wait 3 more months to release scala-pickling.

Finally this approach has been discussed extensively on reflection meetings, together with Adriaan and Jason. We have arrived at the conclusion that it's a good idea, and it should be included into 2.10.x.

**Comment by xeno-by** [comment]
also /cc @adriaanm

**Comment by xeno-by** [comment]
Recent test failures here have nothing to do with this pull request. They are caused by the fact that 2.10.x is currently broken: http://groups.google.com/group/scala-internals/browse_thread/thread/9f1988cf541c6b7d.

Now I fixed the breakage, submitted a fix as a pull request (https://github.com/scala/scala/pull/2513) and rebased this one on top of the fix.

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -96,6 +96,31 @@ trait Implicits {
...
+    }
+    try {
+      wrapper(inferImplicit(tree, pt, reportAmbiguous = true, isView = isView, context = context, saveAmbiguousDivergent = !silent, pos = pos)) match {
+        case failure if failure.tree.isEmpty => fail(None)
+        case success => success.tree
+      }
+    } catch {
+      case ex: DivergentImplicit =>
+        if (settings.Xdivergence211.value)
+          debuglog("this shouldn't happen. DivergentImplicit exception has been thrown with -Xdivergence211 turned on: "+ex)
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: "debuglog" is like "trace". This should be at the other extreme of logging, which is "devWarning" in 2.11 and that should just be backported so we don't have to maintain a distinction, but in 2.10 there is at least "debugwarn" which I realize is confusingly similar in name which may be why it's not called that anymore.

**Comment by retronym** [comment]
LGTM. This is a big improvement, and looks to be carefully enacted. You've already addressed my comments from last week.

**Comment by paulp** [comment]
LGTM2. Except for the fact that it will crash into half a dozen works-in-progress of mine, but such is the life we lead.

**Comment by xeno-by** [comment]
Thank you! I will do the requested minor changes and push them shortly.

**Comment by paulp** [comment]
Feel free to throw in a backport of devWarning (it hardly merits the word "backport", it's just a line of code.)

**Pull request merged**